### PR TITLE
Instrument substream with metrics

### PR DESCRIPTION
**Pull request by thomaseizinger**
As a nice benefit, wrapping the original `Negotiated<yamux::Stream>`
makes the API of `libp2p-xtra` more contained by not exposing details
about `yamux`.

### CODE REVIEW

File: libp2p-xtra/src/substream.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+///
+/// Opening a new substream is cheap compared to say a TCP connection. Additionally, because
+/// substreams are opened on top of an existing connection, they are not affected by NAT or
+/// firewalls. Once a connection is established, either end of the connection (the original dialer
+/// and listener) can open substreams.
+///
+/// Each substream is dedicated to a specific protocol which must be specified upon construction.
+///
+/// Substreams are instrumented with prometheus metrics that track the duration they are alive for
+/// and how many bytes are read from and written to the stream.
```
Comment (luckysori) [inline libp2p-xtra/src/substream.rs:29]: 👍 Thanks for the detailed docstring.

File: libp2p-xtra/src/substream.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        let this = self.project();
+
+        this.inner.poll_close(cx)
+    }
+}
+
+const PROTOCOL_LABEL: &str = "protocol";
```
Comment (luckysori) [inline libp2p-xtra/src/substream.rs:134]: ❓ Does this mean that every protocol will have the same label? Shouldn't this be configurable?

**Review by luckysori — APPROVED** [review]
I love this idea and how quickly you implemented it!

File: libp2p-xtra/src/substream.rs (hunk 2 above)
Comment (thomaseizinger) [reply libp2p-xtra/src/substream.rs:134]: This is the label "key"! The labels are a `HashMap`. One label can have as many values (the actual protocol names) as you want but you need to specify the labels upfront!

**Comment by thomaseizinger** [comment]
> I love this idea and how quickly you implemented it!

Thanks! I am quite stoked myself about what we can do with this data :)

**Comment by thomaseizinger** [comment]
Note to self: Include dialer/listener role in metric.

**Review by klochowicz — APPROVED** [review]
really cool 👍

**Pull request merged**